README.md
# Lolcommits HipChat

[![Gem](https://img.shields.io/gem/v/lolcommits-hipchat.svg?style=flat)](http://rubygems.org/gems/lolcommits-hipchat)
[![Travis](https://travis-ci.org/lolcommits/lolcommits-hipchat.svg?branch=master)](https://travis-ci.org/lolcommits/lolcommits-hipchat)
[![Depfu](https://img.shields.io/depfu/lolcommits/lolcommits-hipchat.svg?style=flat)](https://depfu.com/github/lolcommits/lolcommits-hipchat)
[![Maintainability](https://api.codeclimate.com/v1/badges/d11d0fa605657e46dc23/maintainability)](https://codeclimate.com/github/lolcommits/lolcommits-hipchat/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/d11d0fa605657e46dc23/test_coverage)](https://codeclimate.com/github/lolcommits/lolcommits-hipchat/test_coverage)

[lolcommits](https://lolcommits.github.io/) takes a snapshot with your webcam
every time you git commit code, and archives a lolcat style image with it. Git
blame has never been so much fun!

This plugin shares lolcommit images to a HipChat room, along with a randomized
message with the commit SHA. Your lolcommit will appear like this (perhaps
without the horse):

![example
commit](https://github.com/lolcommits/lolcommits-hipchat/raw/master/assets/images/example-commit.png)


## Requirements

* [Lolcommits](https://lolcommits.github.io/) >= 0.9.5
* Ruby >= 2.0.0
* A webcam
* [ImageMagick](http://www.imagemagick.org)
* [ffmpeg](https://www.ffmpeg.org) (optional) for animated gif capturing

## Installation

After installing the lolcommits gem, install this plugin with:

    $ gem install lolcommits-hipchat

Visit `https://your-team.hipchat.com/account/api` to create a new API token with
the 'Send Message' scope set. Then configure the plugin to enable it:

    $ lolcommits --config -p hipchat
    # set enabled to `true`
    # when prompted, enter your HipChat team name, API token and room (name or id)

That's it! The next lolcommit will be shared to your HipChat room. To disable,
uninstall this gem or use:

    $ lolcommits --config -p hipchat
    # and set enabled to `false`

## Development

Check out this repo and run `bin/setup`, this will install all dependencies and
generate docs. Use `bundle exec rake` to run all tests and generate a coverage
report.

You can also run `bin/console` for an interactive prompt that will allow you to
experiment with the gem code.

## Tests

MiniTest is used for testing. Run the test suite with:

    $ rake test

## Docs

Generate docs for this gem with:

    $ rake rdoc

## Troubles?

If you think something is broken or missing, please raise a new
[issue](https://github.com/lolcommits/lolcommits-hipchat/issues). Take
a moment to check it hasn't been raised in the past (and possibly closed).

## Contributing

Bug [reports](https://github.com/lolcommits/lolcommits-hipchat/issues) and [pull
requests](https://github.com/lolcommits/lolcommits-hipchat/pulls) are welcome on
GitHub.

When submitting pull requests, remember to add tests covering any new behaviour,
and ensure all tests are passing on [Travis
CI](https://travis-ci.org/lolcommits/lolcommits-hipchat). Read the
[contributing
guidelines](https://github.com/lolcommits/lolcommits-hipchat/blob/master/CONTRIBUTING.md)
for more details.

This project is intended to be a safe, welcoming space for collaboration, and
contributors are expected to adhere to the [Contributor
Covenant](http://contributor-covenant.org) code of conduct. See
[here](https://github.com/lolcommits/lolcommits-hipchat/blob/master/CODE_OF_CONDUCT.md)
for more details.

## License

The gem is available as open source under the terms of
[LGPL-3](https://opensource.org/licenses/LGPL-3.0).

## Links

* [Travis CI](https://travis-ci.org/lolcommits/lolcommits-hipchat)
* [Code Climate](https://codeclimate.com/github/lolcommits/lolcommits-hipchat)
* [Test Coverage](https://codeclimate.com/github/lolcommits/lolcommits-hipchat/coverage)
* [RDoc](http://rdoc.info/projects/lolcommits/lolcommits-hipchat)
* [Issues](http://github.com/lolcommits/lolcommits-hipchat/issues)
* [Report a bug](http://github.com/lolcommits/lolcommits-hipchat/issues/new)
* [Gem](http://rubygems.org/gems/lolcommits-hipchat)
* [GitHub](https://github.com/lolcommits/lolcommits-hipchat)


CHANGELOG.md
# ChangeLog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](KeepAChangelog) and this project
adheres to [Semantic Versioning](Semver).

## [Unreleased]

- Your contribution here!

## [0.1.1] - 2018-05-24
### Changed
- Updated gemspec meta data links.

## [0.1.0] - 2018-03-27
### Changed
- Update initializer to accept plugin name string arg
- Require at least lolcommits 0.12.0

### Removed
- runner_order method

## [0.0.7] - 2018-02-05
### Changed
- Require at least lolcommits 0.11.0

### Removed
- Dropped support for Ruby 2.0

## [0.0.6] - 2017-01-13
### Changed
- Update Travis Rubies

### Removed
- Remove name method

## [0.0.5] - 2017-12-22
### Changed
- Rename HipChat to Hipchat everywhere

## [0.0.4] - 2018-12-06

## [0.0.3] - 2017-12-06
### Changed
- Require mime-types gem
- Update README

## [0.0.2] - 2017-10-29
### Changed
- README

## [0.0.1] - 2017-10-26
### Changed
- Initial release

[Unreleased]: https://github.com/lolcommits/lolcommits-hipchat/compare/v0.1.1...HEAD
[0.1.1]: https://github.com/lolcommits/lolcommits-hipchat/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/lolcommits/lolcommits-hipchat/compare/v0.0.7...v0.1.0
[0.0.7]: https://github.com/lolcommits/lolcommits-hipchat/compare/v0.0.6...v0.0.7
[0.0.6]: https://github.com/lolcommits/lolcommits-hipchat/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/lolcommits/lolcommits-hipchat/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/lolcommits/lolcommits-hipchat/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/lolcommits/lolcommits-hipchat/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/lolcommits/lolcommits-hipchat/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/lolcommits/lolcommits-hipchat/compare/b07ddc4...v0.0.1
[KeepAChangelog]: http://keepachangelog.com/en/1.0.0/
[Semver]: http://semver.org/spec/v2.0.0.html


test/test_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

# lolcommits gem
require 'lolcommits'

# lolcommit test helpers
require 'lolcommits/test_helpers/git_repo'
require 'lolcommits/test_helpers/fake_io'

if ENV['COVERAGE']
  require 'simplecov'
end

# plugin gem test libs
require 'lolcommits/hipchat'
require 'minitest/autorun'

# swallow all debug output during test runs
def debug(msg); end


test/lolcommits/plugin/hipchat_test.rb
require "test_helper"
require 'webmock/minitest'

describe Lolcommits::Plugin::Hipchat do

  include Lolcommits::TestHelpers::GitRepo
  include Lolcommits::TestHelpers::FakeIO

  describe "with a runner" do
    def runner
      # a simple lolcommits runner with an empty configuration Hash
      @runner ||= Lolcommits::Runner.new(
        main_image: Tempfile.new('main_image.jpg'),
        config: OpenStruct.new(
          loldir: File.expand_path("#{__dir__}../../../images")
        )
      )
    end

    def plugin
      @plugin ||= Lolcommits::Plugin::Hipchat.new(runner: runner)
    end

    def endpoint
      "http://#{plugin_config[:api_team]}.hipchat.com/v2/room/#{plugin_config[:api_room]}/share/file?auth_token=#{plugin_config[:api_token]}"
    end

    def plugin_config
      {
        enabled: true,
        api_team: "lolcommits-team",
        api_token: "f0FJmP9wfP9JxNeCYrwmSR9f86jJfxgMna1r6mXy",
        api_room: "lolcommits-room"
      }
    end

    describe "#enabled?" do
      it "is disabled by default" do
        plugin.enabled?.must_equal false
      end

      it "is true when configured" do
        plugin.configuration = plugin_config
        plugin.enabled?.must_equal true
      end
    end

    describe "run_capture_ready" do
      before do
        commit_repo_with_message("first commit!")
        plugin.configuration = plugin_config
      end

      after { teardown_repo }

      it "posts lolcommit image and message to HipChat" do
        in_repo do
          stub_request(:post, endpoint).to_return(status: 204)
          output = fake_io_capture { plugin.run_capture_ready }
          output.must_equal "Posting to HipChat (lolcommits-room) ... done!\n"

          assert_requested :post, endpoint, times: 1 do |req|
            req.headers["Content-Type"].must_match(/multipart\/related; boundary=/)
            req.body.must_match(/{\"message\":\"commited some .+ to .+@.+ (.+) \"}/)
          end
        end
      end

      it "prompts user to check token when post to HipChat fails" do
        in_repo do
          stub_request(:post, endpoint).to_return(status: 401)
          output = fake_io_capture { plugin.run_capture_ready }
          output.split("\n").must_equal [
            "Posting to HipChat (lolcommits-room) ... failed!",
            "Are you sure your HipChat API token has the 'Send Message' scope?"
          ]
        end
      end
    end

    describe "configuration" do
      it "allows plugin options to be configured" do
        # enabled, token, team and room
        inputs = %w(
          true
          lolcommits-team
          my-token
          lolcommits-room
        )
        configured_plugin_options = {}

        fake_io_capture(inputs: inputs) do
          configured_plugin_options = plugin.configure_options!
        end

        configured_plugin_options.must_equal({
          enabled: true,
          api_team: "lolcommits-team",
          api_token: "my-token",
          api_room: "lolcommits-room"
        })
      end

      describe "#valid_configuration?" do
        it "returns false for an invalid configuration" do
          plugin.configuration = { api_team: "set-but-other-keys-missing" }
          plugin.valid_configuration?.must_equal false
        end

        it "returns true with a valid configuration" do
          plugin.configuration = plugin_config
          plugin.valid_configuration?.must_equal true
        end
      end
    end
  end
end


test/lolcommits/plugin/hip_chat_test.rb
require "test_helper"
require 'webmock/minitest'

describe Lolcommits::Plugin::HipChat do

  include Lolcommits::TestHelpers::GitRepo
  include Lolcommits::TestHelpers::FakeIO

  def plugin_name
    "hipchat"
  end

  it "should have a name" do
    ::Lolcommits::Plugin::HipChat.name.must_equal plugin_name
  end

  it "should run on capture ready" do
    ::Lolcommits::Plugin::HipChat.runner_order.must_equal [:capture_ready]
  end

  describe "with a runner" do
    def runner
      # a simple lolcommits runner with an empty configuration Hash
      @runner ||= Lolcommits::Runner.new(
        main_image: Tempfile.new('main_image.jpg'),
        config: OpenStruct.new(
          read_configuration: {},
          loldir: File.expand_path("#{__dir__}../../../images")
        )
      )
    end

    def plugin
      @plugin ||= Lolcommits::Plugin::HipChat.new(runner: runner)
    end

    def endpoint
      config = plugin.config.read_configuration['hipchat']
      "http://#{config['api_team']}.hipchat.com/v2/room/#{config['api_room']}/share/file?auth_token=#{config['api_token']}"
    end

    def valid_enabled_config
      @config ||= OpenStruct.new(
        read_configuration: {
          "hipchat" => {
            "enabled"   => true,
            "api_team"  => "lolcommits-team",
            "api_token" => "f0FJmP9wfP9JxNeCYrwmSR9f86jJfxgMna1r6mXy",
            "api_room"  => "lolcommits-room"
          }
        }
      )
    end

    describe "#enabled?" do
      it "is false by default" do
        plugin.enabled?.must_equal false
      end

      it "is true when configured" do
        plugin.config = valid_enabled_config
        plugin.enabled?.must_equal true
      end
    end

    describe "run_capture_ready" do
      before do
        commit_repo_with_message("first commit!")
        plugin.config = valid_enabled_config
      end

      after { teardown_repo }

      it "posts lolcommit image and message to HipChat" do
        in_repo do
          stub_request(:post, endpoint).to_return(status: 204)
          output = fake_io_capture { plugin.run_capture_ready }
          output.must_equal "Posting to HipChat (lolcommits-room) ... done!\n"

          assert_requested :post, endpoint, times: 1 do |req|
            req.headers["Content-Type"].must_match(/multipart\/related; boundary=/)
            req.body.must_match(/{\"message\":\"commited some .+ to .+@.+ (.+) \"}/)
          end
        end
      end

      it "prompts user to check token when post to HipChat fails" do
        in_repo do
          stub_request(:post, endpoint).to_return(status: 401)
          output = fake_io_capture { plugin.run_capture_ready }
          output.split("\n").must_equal [
            "Posting to HipChat (lolcommits-room) ... failed!",
            "Are you sure your HipChat API token has the 'Send Message' scope?"
          ]
        end
      end
    end

    describe "configuration" do
      it "returns false when not configured" do
        plugin.configured?.must_equal false
      end

      it "returns true when configured" do
        plugin.config = valid_enabled_config
        plugin.configured?.must_equal true
      end

      it "allows plugin options to be configured" do
        # enabled, token, team and room
        inputs = %w(
          true
          lolcommits-team
          my-token
          lolcommits-room
        )
        configured_plugin_options = {}

        fake_io_capture(inputs: inputs) do
          configured_plugin_options = plugin.configure_options!
        end

        configured_plugin_options.must_equal({
          "enabled"   => true,
          "api_team"  => "lolcommits-team",
          "api_token" => "my-token",
          "api_room"  => "lolcommits-room"
        })
      end

      describe "#valid_configuration?" do
        it "returns false for an invalid configuration" do
          plugin.config = OpenStruct.new(read_configuration: {
            "hipchat" => { "api_team" => "set-but-other-keys-missing" }
          })
          plugin.valid_configuration?.must_equal false
        end

        it "returns true with a valid configuration" do
          plugin.config = valid_enabled_config
          plugin.valid_configuration?.must_equal true
        end
      end
    end
  end
end


lib/lolcommits/plugin/hipchat.rb
require 'mime/types'
require 'net/http'
require 'lolcommits/plugin/base'

module Lolcommits
  module Plugin
    class Hipchat < Base

      ##
      # Returns true/false indicating if the plugin has been correctly
      # configured. To post a message to HipChat all plugin options must be
      # set.
      #
      # @return [Boolean] true/false indicating if plugin is correctly
      # configured
      #
      def valid_configuration?
        [:api_token, :api_team, :api_room].all? do |option|
          !configuration[option].to_s.strip.empty?
        end
      end

      ##
      # Prompts the user to configure plugin options.
      # Options are enabled (true/false), HipChat API token, team name and room
      # id (or name)
      #
      # @return [Hash] a hash of the configured plugin options
      #
      def configure_options!
        options = super
        options.merge!(configure_auth_options) if options[:enabled]
        options
      end

      ##
      # Post-capture hook, runs after lolcommits captures a snapshot. Shares the
      # lolcommit to the configured HipChat room via their API (v2) using
      # Net::HTTP (POST) with a message and image attachment. See the HipChat
      # API documentation for more information:
      #
      # https://www.hipchat.com/docs/apiv2/method/share_file_with_room
      #
      # @return [Boolean] indicating if HipChat post was successful
      #
      def run_capture_ready
        print "Posting to HipChat (#{configuration[:api_room]}) ..."
        debug "Posting to HipChat: HTTP POST to #{api_url}"

        boundary = "0123456789ABLEWASIEREISAWELBA9876543210"
        http     = Net::HTTP.new(api_url.host)
        header   = { 'Content-Type' => "multipart/related; boundary=#{boundary}" }
        data     = [message_part, picture_part].map do |part|
          "--#{boundary}\r\n#{part}"
        end.join << "--#{boundary}--"

        response = http.post("#{api_url.path}?#{api_url.query}", data, header)

        if response.code == "204"
          print " done!\n"
          return true
        else
          debug "Posting to HipChat failed with response code #{response.code}"
          print " failed!\n"
        end

        if response.code ==  "401"
          puts "Are you sure your HipChat API token has the 'Send Message' scope?"
        end

        false
      end


      private

      def configure_auth_options
        puts "\n"
        puts '-' * 41
        puts ' Lolcommits HipChat Plugin Configuration'
        puts '-' * 41

        puts "\n1. What is your Team Name (e.g. teamname.hipchat.com)"
        print "> "
        team = parse_user_input(gets.strip)
        puts "\n2. We need a HipChat token (visit https://#{team}.hipchat.com/account/api)"
        puts "(ensure the scope 'Send Message' is selected)"
        print "> "
        token = parse_user_input(gets.strip)
        puts "\n3. Which Room should be we post to? (an id or name)"
        print "> "
        room = parse_user_input(gets.strip)

        {
          api_team: team,
          api_token: token,
          api_room: room
        }
      end

      def message_part
        [
          'Content-Type: application/json; charset=UTF-8',
          'Content-Disposition: attachment; name="metadata"',
          '',
          message_json,
          ''
        ].join "\r\n"
      end

      def message_json
        { message: message }.to_json.force_encoding('utf-8')
      end

      def picture_part
        picture   = File.new(runner.main_image)
        mime_type = MIME::Types.type_for(picture.path)[0] || MIME::Types['application/octet-stream'][0]
        [
          format('Content-Type: %s', mime_type.simplified),
          format('Content-Disposition: attachment; name="file"; filename="%s"', picture.path),
          '',
          "#{picture.read} ",
          ''
        ].join "\r\n"
      end

      def api_url
        URI("http://#{configuration[:api_team]}.hipchat.com/v2/room/#{configuration[:api_room]}/share/file?auth_token=#{configuration[:api_token]}")
      end

      def message
        "commited some #{random_adjective} #{random_object} to #{runner.vcs_info.repo}@#{runner.sha} (#{runner.vcs_info.branch}) "
      end

      def random_object
        %w(screws bolts exceptions errors cookies).sample
      end

      def random_adjective
        adjectives = %w(adaptable adventurous affable affectionate agreeable ambitious amiable amicable amusing brave \
                        bright broad-minded calm careful charming communicative compassionate conscientious considerate \
                        convivial courageous courteous creative decisive determined diligent diplomatic discreet dynamic \
                        easygoing emotional energetic enthusiastic exuberant fair-minded faithful fearless forceful \
                        frank friendly funny generous gentle good gregarious hard-working helpful honest humorous \
                        imaginative impartial independent intellectual intelligent intuitive inventive kind loving loyal \
                        modest neat nice optimistic passionate patient persistent pioneering philosophical placid plucky \
                        polite powerful practical pro-active quick-witted quiet rational reliable reserved resourceful \
                        romantic self-confident self-disciplined sensible sensitive shy sincere sociable straightforward \
                        sympathetic thoughtful tidy tough unassuming understanding versatile warmhearted willing witty)
        adjectives.sample
      end
    end
  end
end


lib/lolcommits/plugin/hip_chat.rb
require 'mime/types'
require 'lolcommits/plugin/base'

module Lolcommits
  module Plugin
    class HipChat < Base

      ##
      # Returns the name of this plugin.
      #
      # @return [String] the plugin name
      #
      def self.name
        'hipchat'
      end

      ##
      # Returns position(s) of when this plugin should run during the capture
      # process. A new message is posted to HipChat when the capture is ready.
      #
      # @return [Array] the position(s) (:capture_ready)
      #
      def self.runner_order
        [:capture_ready]
      end

      ##
      # Returns true if the plugin has been configured. An API token, team name
      # and HipChat room id (or name) are required.
      #
      # @return [Boolean] true/false indicating if plugin is configured
      #
      def configured?
        super &&
          !!(configuration['api_token'] &&
             configuration['api_team'] &&
             configuration['api_room'])
      end

      ##
      # Returns true/false indicating if the plugin has been correctly
      # configured. To post a message to HipChat all plugin options must be
      # set.
      #
      # @return [Boolean] true/false indicating if plugin is correctly
      # configured
      #
      def valid_configuration?
        %w(api_token api_team api_room).all? do |option|
          !configuration[option].to_s.strip.empty?
        end
      end

      ##
      # Prompts the user to configure plugin options.
      # Options are enabled (true/false), HipChat API token, team name and room
      # id (or name)
      #
      # @return [Hash] a hash of the configured plugin options
      #
      def configure_options!
        options = super
        options.merge!(configure_auth_options) if options['enabled']
        options
      end

      ##
      # Post-capture hook, runs after lolcommits captures a snapshot. Shares the
      # lolcommit to the configured HipChat room via their API (v2) using
      # Net::HTTP (POST) with a message and image attachment. See the HipChat
      # API documentation for more information:
      #
      # https://www.hipchat.com/docs/apiv2/method/share_file_with_room
      #
      # @return [Boolean] indicating if HipChat post was successful
      #
      def run_capture_ready
        print "Posting to HipChat (#{configuration['api_room']}) ..."
        debug "Posting to HipChat: HTTP POST to #{api_url}"

        boundary = "0123456789ABLEWASIEREISAWELBA9876543210"
        http     = Net::HTTP.new(api_url.host)
        header   = { 'Content-Type' => "multipart/related; boundary=#{boundary}" }
        data     = [message_part, picture_part].map do |part|
          "--#{boundary}\r\n#{part}"
        end.join << "--#{boundary}--"

        response = http.post("#{api_url.path}?#{api_url.query}", data, header)

        if response.code == "204"
          print " done!\n"
          return true
        else
          debug "Posting to HipChat failed with response code #{response.code}"
          print " failed!\n"
        end

        if response.code ==  "401"
          puts "Are you sure your HipChat API token has the 'Send Message' scope?"
        end

        false
      end


      private

      def configure_auth_options
        puts "\n"
        puts '-' * 50
        puts ' Lolcommits HipChat Plugin Configuration'
        puts '-' * 50

        puts "\n1. What is your Team Name (e.g. teamname.hipchat.com)"
        print "> "
        teamname = parse_user_input(gets.strip)
        puts "\n2. We need a HipChat token (visit https://#{teamname}.hipchat.com/account/api)"
        puts "(ensure the scope 'Send Message' is selected)"
        print "> "
        token = parse_user_input(gets.strip)
        puts "\n3. Which Room should be we post to? (an id or name)"
        print "> "
        room = parse_user_input(gets.strip)

        {
          'api_team'  => teamname,
          'api_token' => token,
          'api_room'  => room
        }
      end

      def message_part
        [
          'Content-Type: application/json; charset=UTF-8',
          'Content-Disposition: attachment; name="metadata"',
          '',
          message_json,
          ''
        ].join "\r\n"
      end

      def message_json
        { message: message }.to_json.force_encoding('utf-8')
      end

      def picture_part
        picture   = File.new(runner.main_image)
        mime_type = MIME::Types.type_for(picture.path)[0] || MIME::Types['application/octet-stream'][0]
        [
          format('Content-Type: %s', mime_type.simplified),
          format('Content-Disposition: attachment; name="file"; filename="%s"', picture.path),
          '',
          "#{picture.read} ",
          ''
        ].join "\r\n"
      end

      def api_url
        URI("http://#{configuration['api_team']}.hipchat.com/v2/room/#{configuration['api_room']}/share/file?auth_token=#{configuration['api_token']}")
      end

      def message
        "commited some #{random_adjective} #{random_object} to #{runner.vcs_info.repo}@#{runner.sha} (#{runner.vcs_info.branch}) "
      end

      def random_object
        %w(screws bolts exceptions errors cookies).sample
      end

      def random_adjective
        adjectives = %w(adaptable adventurous affable affectionate agreeable ambitious amiable amicable amusing brave \
                        bright broad-minded calm careful charming communicative compassionate conscientious considerate \
                        convivial courageous courteous creative decisive determined diligent diplomatic discreet dynamic \
                        easygoing emotional energetic enthusiastic exuberant fair-minded faithful fearless forceful \
                        frank friendly funny generous gentle good gregarious hard-working helpful honest humorous \
                        imaginative impartial independent intellectual intelligent intuitive inventive kind loving loyal \
                        modest neat nice optimistic passionate patient persistent pioneering philosophical placid plucky \
                        polite powerful practical pro-active quick-witted quiet rational reliable reserved resourceful \
                        romantic self-confident self-disciplined sensible sensitive shy sincere sociable straightforward \
                        sympathetic thoughtful tidy tough unassuming understanding versatile warmhearted willing witty)
        adjectives.sample
      end
    end
  end
end
